datadog_checks_base/datadog_checks/base/utils/db/utils.py

```diff
@@ -259,6 +262,10 @@ def cancel(self):
         Send a signal to cancel the job loop asynchronously.
         """
         self._cancel_event.set()
+        # after setting cancel event, wait for job loop to fully shutdown
+        if self._job_loop_future:
+            while self._job_loop_future.running():
+                time.sleep(0.1)
```

Comment:
without this change, we see a race condition that can happen when multiple job loop threads are running & `cancel` is called. This can sometimes result in a segfault: 

```
tests/test_statements.py::test_async_job_enabled[True3-True1-True0] Fatal Python error: Segmentation fault

Current thread 0x00007f4924d0b700 (most recent call first):
  File "/home/ec2-user/.local/share/hatch/env/virtual/datadog-postgres/550Gv7FF/py3.9-12.1/lib/python3.9/site-packages/psycopg/cursor.py", line 516 in _select_current_result
  File "/home/ec2-user/.local/share/hatch/env/virtual/datadog-postgres/550Gv7FF/py3.9-12.1/lib/python3.9/site-packages/psycopg/cursor.py", line 205 in _execute_gen
  File "/home/ec2-user/.local/share/hatch/env/virtual/datadog-postgres/550Gv7FF/py3.9-12.1/lib/python3.9/site-packages/psycopg/connection.py", line 957 in wait
....
...
  File "/home/ec2-user/dd/integrations-core/datadog_checks_base/datadog_checks/base/utils/tracking.py", line 71 in wrapper
  File "/home/ec2-user/dd/integrations-core/postgres/datadog_checks/postgres/statement_samples.py", line 434 in run_job
  File "/home/ec2-user/dd/integrations-core/datadog_checks_base/datadog_checks/base/utils/db/utils.py", line 348 in _run_job_traced
  File "/home/ec2-user/dd/integrations-core/datadog_checks_base/datadog_checks/base/utils/db/utils.py", line 342 in _run_job_rate_limited
  File "/home/ec2-user/dd/integrations-core/datadog_checks_base/datadog_checks/base/utils/db/utils.py", line 303 in _job_loop
  File "/home/ec2-user/anaconda3/envs/agent39/lib/python3.9/concurrent/futures/thread.py", line 58 in run
  File "/home/ec2-user/anaconda3/envs/agent39/lib/python3.9/concurrent/futures/thread.py", line 83 in _worker
  ```
  
  This is because the call to `cancel` would simply set the cancel() threading event & would return immediately, which would result in a call to `close_all_connections`. This is a problem bc setting the `cancel` event on the job loop thread does not guarantee that the thread isn't _already executing_ at the time we end up calling close for all connections in the db pool. (we only check if cancel is set [here](https://github.com/DataDog/integrations-core/blob/master/datadog_checks_base/datadog_checks/base/utils/db/utils.py#L290) and [here](https://github.com/DataDog/integrations-core/blob/master/datadog_checks_base/datadog_checks/base/utils/db/utils.py#L340))
  
 Therefore, the only way to safely cancel all job loops is to do the following:
 
 1. `cancel` is called from main check thread
 2. Wait for all job loop `cancel` methods to return 
 3. Job loop method `cancel` methods do not return until the job loop is no longer running 
 4. Once all job loop thread `cancel`'s return, close all database connections in the conn pool
 5. Finally, close main check connection once check is no longer running